### ISSUE TITLE
FSDP does not reduce the gradient size during backward

### ISSUE THREAD

**Issue by gaow0007** [pytorch/pytorch#160320]
### 🐛 Describe the bug

I am currently working with Fully Sharded Data Parallel (FSDP) and investigating its internal workflow, particularly focusing on the backward pass. Initially, I expected FSDP to perform a reduce-scatter operation during the backward stage to reduce gradient memory consumption, thereby storing only a shard of the gradients proportional to the world size. However, when examining the implementation in the function at [https://github.com/pytorch/pytorch/blob/v2.6.0/torch/distributed/fsdp/_runtime_utils.py#L691 ](https://github.com/pytorch/pytorch/blob/v2.6.0/torch/distributed/fsdp/_runtime_utils.py#L691), I observed that the gradient size remains equal to the total parameter size—even with an FSDP world size of 2.

 This is out of my expectation that the gradient memory footprint would be approximately half of the total parameter size due to sharding.

## Reproduce
The reproduce cod is [here.](https://gist.github.com/gaow0007/ce20eeb0cbb5ca081ede846212e13911)
The running command is ```torchrun --nproc_per_node=2 --master_port=29501 test_fsdp_bp.py 1 False```

The first argument is to create a FSDP model with the FSDP size as 2. 
The second argument is to disable model.no_sync behavior. 

## Log
```
W0811 23:03:54.612000 474111 site-packages/torch/distributed/run.py:792] 
W0811 23:03:54.612000 474111 site-packages/torch/distributed/run.py:792] *****************************************
W0811 23:03:54.612000 474111 site-packages/torch/distributed/run.py:792] Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed. 
W0811 23:03:54.612000 474111 site-packages/torch/distributed/run.py:792] *****************************************
your are running normal FSDP
your are running normal FSDP
[1] autograd_computed_grad: 88.96498107910156, torch.Size([400])
[1] autograd_computed_grad: 135.11334228515625, torch.Size([400])
[2] autograd_computed_grad: 44.48249053955078, torch.Size([400])
[2] autograd_computed_grad: 67.55667114257812, torch.Size([400])
before optimizer.step is 69.44602966308594, torch.Size([200]) rank 1
before optimizer.step is 36.28430938720703, torch.Size([200]) rank 0
[1] autograd_computed_grad: 48.636356353759766, torch.Size([400])
[1] autograd_computed_grad: 40.29658126831055, torch.Size([400])
[2] autograd_computed_grad: 24.318178176879883, torch.Size([400])
before optimizer.step is 37.45566177368164, torch.Size([200]) rank 0
[2] autograd_computed_grad: 20.148290634155273, torch.Size([400])
before optimizer.step is 94.09107208251953, torch.Size([200]) rank 1
[rank0]:[W811 23:03:59.718225027 ProcessGroupNCCL.cpp:4561] [PG ID 0 PG GUID 0 Rank 0]  using GPU 0 to perform barrier as devices used by this process are currently unknown. This can potentially cause a hang if this rank to GPU mapping is incorrect. Specify device_ids in barrier() to force use of a particular device, or call init_process_group() with a device_id.
[rank1]:[W811 23:03:59.718420844 ProcessGroupNCCL.cpp:4561] [PG ID 0 PG GUID 0 Rank 1]  using GPU 1 to perform barrier as devices used by this process are currently unknown. This can potentially cause a hang if this rank to GPU mapping is incorrect. Specify device_ids in barrier() to force use of a particular device, or call init_process_group() with a device_id.
# device_mesh = DeviceMesh('cuda', [[0, 1]], mesh_dim_names=('ddp', 'fsdp')), ident = [0.15925315022468567, 0.12011697143316269]
use_no_sync is  False
[Rank 0] Full sync test passed!
use_no_sync is  False
[Rank 1] Full sync test passed!
```

## Expected Behavior
The model has a total of 400 parameters, so we expect each GPU to store a gradient shard of size 200. Therefore, the line starting with ```[1] autograd_computed_grad``` should report ```torch.Size([200])``` rather than ```torch.Size([400])```.

### Versions

Collecting environment information...
PyTorch version: 2.6.0+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.30.2
Libc version: glibc-2.35

Python version: 3.10.18 (main, Jun  5 2025, 13:14:17) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-1040-nvidia-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.6.20
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA H800
GPU 1: NVIDIA H800

Nvidia driver version: 535.129.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.3.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      52 bits physical, 57 bits virtual
Byte Order:                         Little Endian
CPU(s):                             224
On-line CPU(s) list:                0-223
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) Platinum 8480CL
CPU family:                         6
Model:                              143
Thread(s) per core:                 2
Core(s) per socket:                 56
Socket(s):                          2
Stepping:                           7
CPU max MHz:                        3800.0000
CPU min MHz:                        800.0000
BogoMIPS:                           4000.00
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 invpcid_single intel_ppin cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities
L1d cache:                          5.3 MiB (112 instances)
L1i cache:                          3.5 MiB (112 instances)
L2 cache:                           224 MiB (112 instances)
L3 cache:                           210 MiB (2 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-55,112-167
NUMA node1 CPU(s):                  56-111,168-223
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-cusparselt-cu12==0.6.2
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] torch==2.6.0
[pip3] torchaudio==2.6.0
[pip3] torchprofile==0.0.4
[pip3] torchvision==0.21.0
[pip3] transformer_engine_torch==2.2.0
[pip3] triton==3.2.0
[conda] numpy                                       1.26.4           pypi_0           pypi
[conda] nvidia-cublas-cu12                          12.4.5.8         pypi_0           pypi
[conda] nvidia-cuda-cupti-cu12                      12.4.127         pypi_0           pypi
[conda] nvidia-cuda-nvrtc-cu12                      12.4.127         pypi_0           pypi
[conda] nvidia-cuda-runtime-cu12                    12.4.127         pypi_0           pypi
[conda] nvidia-cudnn-cu12                           9.1.0.70         pypi_0           pypi
[conda] nvidia-cufft-cu12                           11.2.1.3         pypi_0           pypi
[conda] nvidia-curand-cu12                          10.3.5.147       pypi_0           pypi
[conda] nvidia-cusolver-cu12                        11.6.1.9         pypi_0           pypi
[conda] nvidia-cusparse-cu12                        12.3.1.170       pypi_0           pypi
[conda] nvidia-cusparselt-cu12                      0.6.2            pypi_0           pypi
[conda] nvidia-nccl-cu12                            2.21.5           pypi_0           pypi
[conda] nvidia-nvjitlink-cu12                       12.4.127         pypi_0           pypi
[conda] nvidia-nvtx-cu12                            12.4.127         pypi_0           pypi
[conda] torch                                       2.6.0            pypi_0           pypi
[conda] torchaudio                                  2.6.0            pypi_0           pypi
[conda] torchprofile                                0.0.4            pypi_0           pypi
[conda] torchvision                                 0.21.0           pypi_0           pypi
[conda] transformer-engine-torch                    2.2.0            pypi_0           pypi
[conda] triton                                      3.2.0            pypi_0           pypi

**Comment by gaow0007** [comment]
One more question. Despite currently PyTorch increases memory usage of FSDP model when we enable [no_sync mode](https://docs.pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.FullyShardedDataParallel.no_sync).

If I want small memory usage even with model.no_sync. 
Could you please advise which files I should focus on in order to implement this functionality?

**Comment by weifengpy** [comment]
Hi @gaow0007 , thanks for the insighful feedback!

> One more question. Currently PyTorch increases memory usage of FSDP model when we enable [no_sync mode](https://docs.pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.FullyShardedDataParallel.no_sync).

the memory increase becaues fsdp1 hold unsharded gradients. if you want to decrease memory usage under no_sync, I can think of two ways
* use cpu offloading. this only helps if memory peaks in non-backward (eg loss computation)
* reduce-scatter gradients, which is equivalent to not using no_sync 

>  observed that the gradient size remains equal to the total parameter size—even with an FSDP world size of 2.

this is not expected. probably due to some config issues

**Comment by gaow0007** [comment]
@weifengpy  Thanks for your explanation. 
I want to make my second question more clear because I have attempted many configs and still do not see the gradient size is the half of the total parameter size. 


## Description
When using FSDP with a world size of 2 and without enabling no_sync mode, I observed the following behavior:

After calling loss.backward(), the gradient size on each device is approximately half of the total parameter size—consistent with FSDP’s sharding strategy, where parameters and gradients are sharded across devices.

However, within the ``_post_backward_hook function`` ([source](https://github.com/pytorch/pytorch/blob/v2.6.0/torch/distributed/fsdp/_runtime_utils.py#L691)), the gradients appear to have the full (total) parameter size before the hook executes. After the hook runs, the gradient size is reduced back to half (i.e., the size of the local shard).

This raises a question: Is the gradient size temporarily restored to the full parameter size during _post_backward_hook, and if so, is this expected behavior? 

I would like to confirm whether gradients are indeed unsharded or reconstructed temporarily in this hook before being reduced back to their sharded form.

Best

**Comment by aditvenk** [comment]
Doing scrubbing -- looks like the question has been clarified. Please re-open the issue or file a new one if you have follow-up questions.
